repository/src/main/java/org/apache/atlas/discovery/EntityLineageService.java

```diff
@@ -454,6 +456,15 @@ private boolean incrementAndCheckIfRelationsLimitReached(AtlasEdge atlasEdge, bo
         String                     outGuid                   = AtlasGraphUtilsV2.getIdFromVertex(outVertex);
         LineageOnDemandConstraints outGuidLineageConstraints = getAndValidateLineageConstraintsByGuid(outGuid, atlasLineageOnDemandContext);
 
+        boolean lineageContainsVisitedEdge = lineageContainsVisitedEdgeV2(ret, atlasEdge);
+        boolean isOutput = AtlasLineageOnDemandInfo.LineageDirection.OUTPUT.equals(direction);
+        boolean isOutVertexBaseVertex = baseGuid.equals(outGuid);
+        boolean isCyclic = lineageContainsVisitedEdge && isOutput && isOutVertexBaseVertex;
```

Comment:
In case of the first edge that is from the base asset to the process asset, (including non-cyclic relations) outGuid will be the base asset. Therefore, if we visited this edge before, it will think this as self-cyclic relation even though it might not be.

To make sure it is self-cyclic, we need also keep track of the outcome of the first edge to second edge if it is directing to base asset. Meaning that, inGuid for OUTPUT direction of the second edge will be referencing base asset.

Reply:
fundamental issue with this approach:

since this method only knows a single particular edge, it can not know the relation all together. We can say if it is OUTPUT direction and inGuid referencing base entity, it is always self-cyclic. But, while processing for the first edge of the self-cyclic relation, this will say this relation is not self-cyclic since inGuid will reference the process vertex. But in fact, it is self-cyclic relation. Therefore, it won't count already added self-cyclic relations and hence the same issue we had after the first fix.

So, it is best to look out for self-cyclic relations in the relations attribute of ret. This will make things much more simpler.